core/android/src/main/kotlin/com/walletconnect/android/internal/common/di/CoreNetworkModule.kt

```diff
@@ -86,16 +83,21 @@ fun coreAndroidNetworkModule(serverUrl: String, connectionType: ConnectionType,
         Interceptor { chain ->
             val request = chain.request()
             try {
-                if (wasFailOvered && request.url.host == DEFAULT_RELAY_URL) {
-                    chain.proceed(request.newBuilder().url(get<String>(named(AndroidCommonDITags.RELAY_URL))).build())
-                } else {
-                    chain.proceed(request)
+                val host = request.url.host
+                when {
+                    shouldFallbackRelay(host) -> chain.proceed(request.newBuilder().url(get<String>(named(AndroidCommonDITags.RELAY_URL))).build())
+                    shouldFallbackEcho(host) -> chain.proceed(request.newBuilder().url(getFallbackEchoUrl(request.url.toString())).build())
+                    shouldFallbackVerify(host) -> chain.proceed(request.newBuilder().url(getFallbackVerifyUrl(request.url.toString())).build())
+                    else -> chain.proceed(request)
                 }
             } catch (e: Exception) {
-                if (request.url.host == DEFAULT_RELAY_URL && isFailOverException(e)) {
-                    SERVER_URL = "$FAIL_OVER_RELAY_URL?projectId=${Uri.parse(SERVER_URL).getQueryParameter("projectId")}"
-                    wasFailOvered = true
-                    chain.proceed(request.newBuilder().url(get<String>(named(AndroidCommonDITags.RELAY_URL))).build())
+                if (isFailOverException(e)) {
+                    when (request.url.host) {
+                        DEFAULT_RELAY_URL.host -> fallbackRelay(chain, request)
+                        DEFAULT_ECHO_URL.host -> fallbackEcho(request, chain)
+                        DEFAULT_VERIFY_URL.host -> fallbackVerify(request, chain)
+                        else -> chain.proceed(request.newBuilder().url(request.url).build())
```

Comment:
we don't need to build a new a request here right? You could just do 
```suggestion
                        else -> chain.proceed(request)
```